### ISSUE TITLE
Editorial: nested surrogate pairs?

### ISSUE THREAD

**Issue by jmdyck** [tc39/ecma262#969]
In the ['Patterns'](https://tc39.github.io/ecma262/#sec-patterns) clause, the Syntax section contains the paragraph:
> Each `\u TrailSurrogate` for which the choice of associated `u LeadSurrogate` is ambiguous shall be associated with the nearest possible `u LeadSurrogate` that would otherwise have no corresponding `\u TrailSurrogate`.

This strikes me as very odd. The wording is exactly parallel to that for resolving the dangling else problem, suggesting that surrogate pairs *nest* somehow. I.e., given
```
\u LeadSurrogate \u LeadSurrogate \u TrailSurrogate \u TrailSurrogate
```
the last `TrailSurrogate` should be 'associated' with the first `LeadSurrogate`. But what would that even mean? (Note that the relevant semantics make no mention of "associated" or "corresponding" Surrogates.)

Now, granted, the grammar *is* formally ambiguous here (`Alternative[+U]` derives `\u LeadSurrogate \u TrailSurrogate` in two distinct ways), and so requires some disambiguation. But rather than the quoted paragraph, I'd prefer one of these approaches:

- If some part of the source text matches `\u LeadSurrogate \u TrailSurrogate`, it must be parsed as a single `Atom` rather than two.
- In cases of ambiguity, RegExpUnicodeEscapeSequence's first RHS is preferred over its second and third.
- Change the second RHS to `[+U] u LeadSurrogate [lookahead != \u TrailSurrogate]`

(One objection to the last approach is that `TrailSurrogate` is not a terminal symbol, and so this is not a 'legal' lookahead sequence. However, this doesn't bother me much: we could enlarge the definition of lookahead sequences if we wanted.)

**Comment by ljharb** [comment]
cc @mathiasbynens

**Comment by mathiasbynens** [comment]
I agree the current paragraph is oddly phrased. I don't understand why it's written like that. Maybe @NorbertLindenberg remembers?

If we're gonna change this, your first suggested approach seems simplest IMHO:

> If some part of the source text matches `\u LeadSurrogate \u TrailSurrogate`, it must be parsed as a single `Atom` rather than two.

**Comment by jmdyck** [comment]
Note that PR #1727 also addresses this issue.

**Comment by NorbertLindenberg** [comment]
I didn’t write this – it first showed up in the spec draft of 2015-04-03. The change was preceded by a discussion on “lonely surrogates and unicode regexps” starting here:
https://mail.mozilla.org/pipermail/es-discuss/2015-January/041281.html

**Comment by allenwb** [comment]
I wrote it because without it (or something like it) the grammar is ambiguous.   The reason the phrasing parallels the language used for "dangling else" clause is because at the grammar engineering level this is the same kind of ambiguity and copying that language seemed like a safe choice to make given this was happening in April 2015 immediately before release of the final ES6 draft.  There wasn't a lot of time to study alternatives.

Note that the disambiguation really only concerns the first three alternative of 

> _RegExpUnicodeEscapeSequence_ ::
    __u__ _LeadSurrogate_ __\u__ _TrailSurrogate_
     __u__ _LeadSurrogate_
      __u__ _TailSurrogate_
      ... 

because that is where the ambiguity exists.  The disambiguation statement does not lead to recognizing "nested surrogate pairs"  in situations such as such as:
>\u LeadSurrogate \u LeadSurrogate \u TrailSurrogate \u TrailSurrogate

 1. In processing the first lead surrogate there is no ambiguity between choosing between alternative  1 and alternative 2 because because the lead surrogate is not immediately followed by a tail surrogate. So, alternative 1 is matched.
 2. In processing the second surrogate there  is an ambiguity. Either alternative 1 or alternative 2 could be used. So, the disambiguation rule applies and forces use of alternative 1, consuming the first tail surrogate.
3. Finally, the second tail surrogate is processed unambiguously matched using alternative 3.

So, no "nesting" is recognized.  

But, I agree that the current wording of the disambiguation statement isn't ideal for this situation. My suggestion for a replacement (I excluded the grammar parameters but the actual spec text should include them):

For ambiguous inputs where either the rule:
_RegExpUnicodeEscapeSequence_ :: __u__ _LeadSurrogate_ __\u__ _TrailSurrogate_ \
or the rule: \
_RegExpUnicodeEscapeSequence_ :: __u__ _LeadSurrogate_ 
could be recognized the first rule always is the one that is recognized.

**Comment by allenwb** [comment]
Attached is the email thread that led to the addition of the paragraph in question.
 Note that it had nothing to do with nesting of surrogate pairs.  Its only intent was to disambiguate the grammar.

[Re ES6 grammar abnormalities.pdf](https://github.com/tc39/ecma262/files/5373826/Re.ES6.grammar.abnormalities.pdf)
